File: source/CSharp/OpenFiniteStateMachine/Java/Core/src/org/ofsm/impl/FileFiniteStateMachine.java
/*
 * Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
 * 
 * Licensed under the LGPL, Version 2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.gnu.org/copyleft/lgpl.html
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * With any your questions welcome to my e-mail 
 * or blog at http://abdulla-a.blogspot.com.
 */

package org.ofsm.impl;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.StreamCorruptedException;

import java.lang.reflect.Type;

import java.net.MalformedURLException;

import java.net.URI;
import java.net.URL;

import java.util.Collection;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.ofsm.IFiniteStateMachineInitiable;
import javax.xml.bind.Unmarshaller;
import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;

import org.ofsm.IEvent;
import org.ofsm.IState;


public class FileFiniteStateMachine extends FiniteStateMachine implements IFiniteStateMachineInitiable{
    private final String machineName = "FileFiniteStateMachine";
    private final String processCriteriaPattern = ".ofsm.xml";
    private URI initByResource;
    private InputStream initByStream;
    private FileFiniteStateMachineModel model = new FileFiniteStateMachineModel();
    
    public String getName() {
        return machineName;        
    }
    
    public boolean mayProcessed(URI params) {
        if(params.getPath().contains(processCriteriaPattern))
            return true;
        else
            return false;
    }

    public boolean mayProcessed(String streamName) {
        if(streamName.contains(processCriteriaPattern))
            return true;
        else
            return false;
    }

    public void init(InputStream stream) throws Exception {
        initByStream = stream;
        unmarshal(initByStream);
    }    

    public synchronized void init(URI params) throws Exception {        
        this.clearTransitions();
        initByResource = params;
        InputStream stream = null;
        if(params.getScheme().equalsIgnoreCase("file")) {
            //String fileName = params.getAuthority() + params.getPath();
            stream = new FileInputStream(new File(params));
        }
        else {
            URL url = initByResource.toURL();
            stream = url.openStream();            
        }
        try {
            unmarshal(stream);
        }
        finally {
            stream.close();
        }
    }
    
    public void reset() throws Exception {
        if(initByResource!=null)
            init(initByResource);
        else
            super.reset();
    }

    private void unmarshal(InputStream stream) throws Exception {
        model.unmarshal(stream);
        setInitialState(model.getInitialState());
        for(Map.Entry<IState, List< StateTransition > > lstItem: model.getTransitionInfos().entrySet()) {
           for(StateTransition item: lstItem.getValue()) {
               addTransition(item);
           }
        }
    
    }
}

File: source/CSharp/OpenFiniteStateMachine/net/org/ofsm/impl/FileFiniteStateMachine.cs
/*
* Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
* 
* Licensed under the LGPL, Version 2 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*      http://www.gnu.org/copyleft/lgpl.html
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* 
* With any your questions welcome to my e-mail 
* or blog at http://abdulla-a.blogspot.com.
*/
using System;
using System.Collections.Generic;
using org.ofsm;

namespace org.ofsm.impl
{		
	[Serializable]
	public class FileFiniteStateMachine:FiniteStateMachine, IFiniteStateMachineInitiable
	{		
		private const string machineName = "FileFiniteStateMachine";		
		private const string processCriteriaPattern = ".ofsm.xml";
		private Uri initByResource;
		private System.IO.Stream initByStream;
		private FileFiniteStateMachineModel model = new FileFiniteStateMachineModel();
		
		public override string getName()
		{
			return machineName;
		}
		
		public virtual bool mayProcessed(Uri location)
		{
            if (location.AbsolutePath.Contains(processCriteriaPattern))
				return true;
			else
				return false;
		}
		
		public virtual bool mayProcessed(string streamName)
		{
			if (streamName.Contains(processCriteriaPattern))
				return true;
			else
				return false;
		}
		
		public virtual void  init(System.IO.Stream stream)
		{
			initByStream = stream;
			unmarshal(initByStream);
		}
				
        public virtual void init(Uri location)
		{
			lock (this)
			{
				this.clearTransitions();
                initByResource = location;
				System.IO.Stream stream = null;                
				if (initByResource.Scheme.Equals("file", StringComparison.OrdinalIgnoreCase))
				{
					string fileName = location.AbsolutePath;
					stream = new System.IO.FileStream(new System.IO.FileInfo(fileName).FullName, System.IO.FileMode.Open, System.IO.FileAccess.Read);
				}
				else
				{
					stream = System.Net.WebRequest.Create(initByResource).GetResponse().GetResponseStream();
				}
				try
				{
					unmarshal(stream);
				}
				finally
				{
                    if(stream!=null)
					    stream.Close();
				}
			}
		}
		
		public override void  reset()
		{
			if (initByResource != null)
				init(initByResource);
			else
				base.reset();
		}
		
		private void  unmarshal(System.IO.Stream stream)
		{
			model.unmarshal(stream);
			setInitialState(model.getInitialState());
            foreach(ICollection<StateTransition> lstItem in model.getTransitionInfos().Values)
			{
				foreach(StateTransition item in lstItem)
				{
					addTransition(item);
				}
			}
		}
	}
}